crates/transport/src/actors/gateway.rs

```diff
@@ -98,72 +198,146 @@ impl GatewayTransport {
+        config.portal_logs_collector_lookup_interval,
+    );
+    (events_rx, handle)
 }
 
 pub struct GatewayBehaviour {
     base: Wrapped<BaseBehaviour>,
+    provider_query: Option<QueryId>,
+    providers_list: Vec<PeerId>,
 }
 
 impl GatewayBehaviour {
     pub fn new(mut base: BaseBehaviour, config: GatewayConfig) -> Wrapped<Self> {
         if config.worker_status_via_gossipsub {
             base.subscribe_heartbeats();
         }
-        if config.worker_status_via_polling {
-            base.start_pulling_heartbeats();
-        }
 
-        Self { base: base.into() }.into()
+        base.start_pulling_heartbeats();
+
+        Self {
+            base: base.into(),
+            provider_query: Default::default(),
+            providers_list: Default::default(),
+        }
+        .into()
     }
 
-    fn on_base_event(&mut self, ev: BaseBehaviourEvent) -> Option<GatewayEvent> {
+    fn on_base_event(&mut self, ev: BaseBehaviourEvent) -> Option<InternalGatewayEvent> {
         match ev {
             BaseBehaviourEvent::Heartbeat { peer_id, heartbeat } => {
-                self.on_heartbeat(peer_id, heartbeat)
+                Some(InternalGatewayEvent::Heartbeat { peer_id, heartbeat })
             }
+            BaseBehaviourEvent::ProviderRecord {
+                id,
+                result,
+                stats,
+                step,
+            } => self.on_provider_record(id, result, stats, step),
             _ => None,
         }
     }
 
-    fn on_heartbeat(&mut self, peer_id: PeerId, heartbeat: Heartbeat) -> Option<GatewayEvent> {
-        log::debug!("Got heartbeat from {peer_id}");
-        log::trace!("{heartbeat:?}");
-        Some(GatewayEvent::Heartbeat { peer_id, heartbeat })
+    fn on_provider_record(
+        &mut self,
+        id: QueryId,
+        result: Result<GetProvidersOk, GetProvidersError>,
+        stats: QueryStats,
+        step: ProgressStep,
+    ) -> Option<InternalGatewayEvent> {
+        if let Some(expected_id) = self.provider_query {
+            if expected_id != id {
+                return None;
+            }
+        } else {
+            return None;
+        }
+        log::debug!("Got provider record: {result:?} {stats:?} {step:?}");
+
+        match result {
+            Ok(GetProvidersOk::FoundProviders { providers, .. }) => {
+                self.providers_list.append(&mut providers.into_iter().collect());
+            }
+            _ => {}
+        }
+        if step.last {
+            self.provider_query = None;
+            return Some(InternalGatewayEvent::LogListeners {
+                listeners: self.providers_list.drain(..).collect()
+            });
+        }
+        None
+    }
+
+    pub fn update_portal_logs_listeners(&mut self) {
+        let key = PORTAL_LOGS_PROVIDER_KEY;
+        if self.provider_query.is_some() {
+            return;
+        }
+        let query_id =
+            self.base.get_kademlia_mut_ref().get_providers(key.to_vec().into());
```

Comment:
I think the `record::Key` can be built directly from `[u8]` without memory allocation for the vector